### PR TITLE
Add tests for retry

### PR DESCRIPTION
**Pull request by kacperkoza**

### CODE REVIEW

File: utils/tests/retry.test.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+const retry = require('../retry');
+const delay = require('../delay');
+const expect = require('chai').expect;
+const sinon = require('sinon');
+
+
+[-100, -1, 0].forEach(invalidRetryCount => {
+    it('should re-throw error from rejected promise when retries count is lower or equal to zero', function (done) {
```
Comment (JeremiAnastaziak) [inline utils/tests/retry.test.js]: serio nie wolisz async await?

Comment (JeremiAnastaziak) [inline utils/tests/retry.test.js]: async await would be much cleaner

File: utils/retry.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module.exports = function retry(fun, options = { timeout: 0, retries: 0 }) {
   return fun().catch((err) => {
-    if (options.retries === 0) {
+    if (options.retries <= 0 || options.timeout < 0) {
```
Comment (JeremiAnastaziak) [inline utils/retry.js]: that moron proof check

File: utils/tests/retry.test.js (hunk 1 above)
Comment (JeremiAnastaziak) [inline utils/tests/retry.test.js]: async await would be much cleaner

File: utils/tests/retry.test.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        const rejectedPromise = () => Promise.reject('invalid input data');
+
+        retry(rejectedPromise, { retries: invalidRetryCount }).catch(err => {
+            expect(err).to.be.equal('invalid input data');
+            done()
+        });
+    });
+});
+
+[-100, -1].forEach(invalidTimeout => {
```
Comment (JeremiAnastaziak) [inline utils/tests/retry.test.js]: whats the difference between -100 and -1? both values are negative
-100 is more negative? 🤣

File: utils/tests/retry.test.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    const spy = sinon.spy(() => Promise.reject('error'));
+
+    retry(spy, { retries: 2, timeout: 10 }).catch(err => {
+            expect(spy.calledThrice).to.be.true;
+            done();
+        }
+    )
+});
+
+it('should retry function after the time specified by timeout value', async function () {
```
Comment (JeremiAnastaziak) [inline utils/tests/retry.test.js]: there you go, async await, much cleaner

File: utils/tests/retry.test.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    await delay(timeout / 2);
+    expect(spy.callCount).to.be.equal(1);
+    await delay(timeout);
+    expect(spy.callCount).to.be.equal(2);
+    await delay(timeout * 2);
+    expect(spy.callCount).to.be.equal(2);
+});
+
+it('should return value if failing promise resolves', function (done) {
```
Comment (JeremiAnastaziak) [inline utils/tests/retry.test.js]: imho it's much cleaner that way:
```
it('returns resolved promise when retries is 5 and source promise is resolved at third call', async () => {
    const calls = [
      () => Promise.reject(err),
      () => Promise.reject(err),
      () => Promise.resolve(success),
    ];
    let count = 0;

    // eslint-disable-next-line no-plusplus
    const fun = sinon.spy(() => calls[count++]());
    await retry(fun, { retries: 5 }).then(resolveSpy);

    expect(fun).to.have.been.calledThrice;
    expect(resolveSpy).to.have.been.calledOnce;
  });
```
also the case you forgot to test

File: utils/tests/retry.test.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
+const retry = require('../retry');
+const delay = require('../delay');
+const expect = require('chai').expect;
+const sinon = require('sinon');
+
+
```
Comment (JeremiAnastaziak) [inline utils/tests/retry.test.js]: could you wrap it in `describe`?

**Review by JeremiAnastaziak — DISMISSED** [review]

File: utils/tests/retry.test.js (hunk 3 above)
Comment (kacperkoza) [reply utils/tests/retry.test.js]: https://www.guru99.com/equivalence-partitioning-boundary-value-analysis.html#1

The basic idea in boundary value testing is to select input variable values at their:
**Minimum
Just above the minimum**
A nominal value
Just below the maximum
Maximum

File: utils/tests/retry.test.js (hunk 6 above)
Comment (kacperkoza) [reply utils/tests/retry.test.js]: +1

File: utils/retry.js (hunk 2 above)
Comment (kacperkoza) [reply utils/retry.js]: ok, 

```
According to the MDN reference, the specification requires that there is a minimum timeout.

If you provide something less than this (HTML5 spec says 4ms) then the browser will just ignore your delay and use the minimum.

So negatives should be fine, since it'll just be less than the minimum.
```
